File: notifications/src/test/java/io/castled/push/trigger/TriggerEvaluatorTest.java
package io.castled.push.trigger;

import static junit.framework.TestCase.assertFalse;
import static junit.framework.TestCase.assertTrue;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import io.castled.notifications.trigger.EventFilterDeserializer;
import io.castled.notifications.trigger.TriggerParamsEvaluator;
import io.castled.notifications.trigger.models.EventFilter;
import io.castled.notifications.trigger.models.NestedEventFilter;

public class TriggerEvaluatorTest {

    @Test
    public void testTriggerEvaluation() {
        TriggerParamsEvaluator triggerParamsEvaluator = new TriggerParamsEvaluator();
        Map<String, Object> trueParams = new HashMap<>();
        trueParams.put("name", "ScreenA");

        Map<String, Object> falseParams = new HashMap<>();
        falseParams.put("name", "ScreenB");

        String triggerExpr = "{\n" +
                "    \"type\": \"NESTED\",\n" +
                "    \"joinType\": \"OR\",\n" +
                "    \"nestedFilters\": [\n" +
                "      {\n" +
                "        \"type\": \"HAVING_PROPERTY\",\n" +
                "        \"name\": \"name\",\n" +
                "        \"operation\": {\n" +
                "          \"type\": \"EQ\",\n" +
                "          \"propertyType\": \"string\",\n" +
                "          \"value\": \"ScreenA\"\n" +
                "        }\n" +
                "      }\n" +
                "    ]\n" +
                "  }";

        Gson gson = new GsonBuilder()
                .registerTypeAdapter(EventFilter.class, new EventFilterDeserializer())
                .create();
        EventFilter eventFilter = gson.fromJson(triggerExpr, EventFilter.class);
        assertTrue(triggerParamsEvaluator.evaluate(trueParams, (NestedEventFilter) eventFilter));
        assertFalse(triggerParamsEvaluator.evaluate(falseParams, (NestedEventFilter) eventFilter));
    }
}

File: notifications/src/test/java/io/castled/notifications/trigger/TriggerEvaluatorTest.kt
package io.castled.notifications.trigger

import io.castled.notifications.trigger.EventFilterEvaluator.evaluate
import io.castled.notifications.trigger.models.*
import io.castled.notifications.workmanager.models.CastledInAppEventRequest
import io.castled.notifications.workmanager.models.CastledNetworkRequest
import io.castled.notifications.workmanager.models.CastledPushEventRequest
import io.castled.notifications.workmanager.models.CastledPushRegisterRequest
import junit.framework.TestCase
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlinx.serialization.modules.SerializersModule
import kotlinx.serialization.modules.polymorphic
import kotlinx.serialization.modules.subclass
import org.junit.Test

class TriggerEvaluatorTest {

    @Test
    fun testTriggerEvaluation() {
        val trueParams: MutableMap<String, Any> = HashMap()
        trueParams["name"] = "ScreenA"
        val falseParams: MutableMap<String, Any> = HashMap()
        falseParams["name"] = "ScreenB"
        val triggerExpr = """{
    "type": "GROUP",
    "joinType": "OR",
    "filters": [
      {
        "type": "HAVING_PROPERTY",
        "name": "name",
        "operation": {
          "type": "EQ",
          "propertyType": "string",
          "value": "ScreenA"
        }
      }
    ]
  }"""

        val eventFilter: EventFilter = Json.decodeFromString(triggerExpr)
        TestCase.assertTrue(evaluate(eventFilter as GroupFilter, trueParams))
        TestCase.assertFalse(evaluate(eventFilter, falseParams))
    }
}